### ISSUE TITLE
[AWSCore] API request signature uses uncoded query parameters

### ISSUE THREAD

**Issue by edwinveger** [aws-amplify/aws-sdk-ios#3362]
This issue is related to / a duplicate of [issue 977 in the Amplify repository](https://github.com/aws-amplify/amplify-ios/issues/977). I have traced the issue to AWSCore.

**Describe the bug**
Signing takes place prior to URL query encoding. This fails the request if there are query parameters contains characters that need to be encoded, e.g. `@` or `:`.

**To Reproduce**
Perform a REST request with a query parameters that contains one or more characters that will be encoded by URLSession  when making the request (e.g. `/my/endpoint/path/?email=my@email.address.com` or `/dates/post/?s=2021-06-18T09:00:00Z`).

**Observed Behavior**
In Amplify this fails with a 403 response status code.

```
{
  "error": "The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.\n\nThe Canonical String for this request should have been\n'GET\n/my/endpoint/path/\nemail=my%40email.address.com\ncontent-type:application/json\nhost:REDACTED\nuser-agent:amplify-iOS/(Function) iOS/14.3 en\nx-amz-date:20201217T105627Z\nx-amz-security-token:REDACTED\n\ncontent-type;host;user-agent;x-amz-date;x-amz-security-token\REDACTED'\n\nThe String-to-Sign should have been\n'AWS4-HMAC-SHA256\REDACTED\n20201217/us-east-1/execute-api/aws4_request\REDACTED'\n"
}
```

**Expected Behavior**
A successful request since signing should be OK.

**Suggested fix**
I have traced the issue to `AWSSignature.m`. If I add the following lines at line 327 (right after setting `query`), it works.

```swift
NSCharacterSet* querySet = [[NSCharacterSet characterSetWithCharactersInString:@"!*'\();:@+$,%#[] "] invertedSet];
query = [query stringByAddingPercentEncodingWithAllowedCharacters:querySet];
```
This ensures that the proper encoded query string is used for signing.
It seems like a very common use case, and seeing the simplicity of this fix I must say I am quite confused that this issue has not been reported/addressed earlier. Am I missing something?

**Misc**
After waiting over three weeks for a response to [the reported issue in the other repository](](https://github.com/aws-amplify/amplify-ios/issues/977)), I decided to pursue a fix myself. Since the fix is rather simple, I'd be happy to submit a PR if I knew that it'd be merged promptly. I would intend to also add a few tests to `AWSSignatureTests`.

**Comment by wooj2** [comment]
Hi @edwinveger ,

Sorry for the delay.  I agree that it is odd that something like this hasn't been caught/reported considering the SDK has been in use for many years now.  Would it be possible to post a PR with some tests so that we can take a closer look?

Thanks!

**Comment by edwinveger** [comment]
Right, I'll look into it (it'll probably be a couple of days).

**Comment by edwinveger** [comment]
Definitely related to #3215.

I have looked into writing tests, but I couldn't find an easy entry point.

I'm proposing a modification to AWSSignatureV4Signer, the method 
`- (NSString *)signRequestV4:(NSMutableURLRequest *)request credentials:(AWSCredentials *)credentials {`. 

I had hoped to add a few testcases to an existing set, but AFAICT there are no tests directly covering this method. When running the tests on the AWSCore target (generated from `xed .` as per instruction) and checking the generated code coverage, it seems like this method is not covered at all 🤔  I'm missing the in-depth knowledge about this framework and the signing algorithm to write tests from scratch.

I've created a meager PR (#3378) indicating the placement and nature of the fix.
